pkg/upgrade/upgrade.go

```diff
@@ -652,3 +662,718 @@ func cleanupDeprecatedKueueVAPB(ctx context.Context, cli client.Client) error {
+	notebooksOnlyToleration, err := getNotebooksOnlyToleration(odhConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get notebooks-only toleration: %w", err)
+	}
+
+	// Calculate container resource limits
+	notebookContainerCounts, err := CalculateContainerResourceLimits(odhConfig, "notebookSizes")
+	if err != nil {
+		return fmt.Errorf("failed to calculate notebook container limits: %w", err)
+	}
+	// default min limits only for serving HWP, max limits are not set
+	servingContainerCounts := map[string]string{
+		"minMemory": "1Gi",
+		"minCpu":    "1",
+	}
+
+	var multiErr *multierror.Error
+
+	// Create 2 HWPs for each AcceleratorProfile
+	for _, ap := range apList.Items {
+		// Create notebooks HWP
+		if err := createHardwareProfileFromAcceleratorProfile(cli, ctx, ap, "notebooks", notebookContainerCounts, notebooksOnlyToleration); err != nil {
+			multiErr = multierror.Append(multiErr, fmt.Errorf("failed to create notebooks HWP for AP %s: %w", ap.GetName(), err))
+			continue
+		}
+
+		// Create serving HWP
+		if err := createHardwareProfileFromAcceleratorProfile(cli, ctx, ap, "serving", servingContainerCounts, nil); err != nil {
+			multiErr = multierror.Append(multiErr, fmt.Errorf("failed to create serving HWP for AP %s: %w", ap.GetName(), err))
+			continue
+		}
+	}
+
+	return multiErr.ErrorOrNil()
+}
+
+func createHardwareProfileFromAcceleratorProfile(cli client.Client, ctx context.Context,
+	ap unstructured.Unstructured, profileType string, containerCounts map[string]string,
+	toleration []corev1.Toleration) error {
+	log := logf.FromContext(ctx)
+	apName := ap.GetName()
+	hwp, err := generateHardwareProfileFromAcceleratorProfile(ap, profileType, containerCounts, toleration)
+	if err != nil {
+		return fmt.Errorf("failed to generate %s HardwareProfile for AcceleratorProfile '%s' (profileType: %s): %w", profileType, apName, profileType, err)
+	}
+
+	if err := cli.Create(ctx, hwp); err != nil {
+		if !k8serr.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create %s HardwareProfile '%s' for AcceleratorProfile '%s' (profileType: %s): %w", profileType, hwp.GetName(), apName, profileType, err)
+		} else {
+			log.Info(fmt.Sprintf("%s HardwareProfile already exists for AcceleratorProfile", profileType),
+				"hardwareProfileName", hwp.GetName(), "acceleratorProfileName", apName, "profileType", profileType)
+		}
+	} else {
+		log.Info(fmt.Sprintf("%s HardwareProfile created for AcceleratorProfile", profileType),
+			"hardwareProfileName", hwp.GetName(), "acceleratorProfileName", apName, "profileType", profileType)
+	}
+	return nil
+}
+
+func getAcceleratorProfiles(ctx context.Context, cli client.Client, applicationNS string) (*unstructured.UnstructuredList, error) {
+	apList := &unstructured.UnstructuredList{}
+	apList.SetGroupVersionKind(gvk.DashboardAcceleratorProfile)
+	err := cli.List(ctx, apList, client.InNamespace(applicationNS))
+	if err != nil {
+		if meta.IsNoMatchError(err) {
+			return nil, nil
+		}
+		return nil, err
```

Comment:
is usually a good practice to wrap the error to help finding the source of the error